sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScanBuilder.scala

```diff
@@ -27,15 +27,20 @@ abstract class FileScanBuilder(
     dataSchema: StructType) extends ScanBuilder with SupportsPushDownRequiredColumns {
   private val partitionSchema = fileIndex.partitionSchema
   private val isCaseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
+  protected val supportsNestedSchemaPruning: Boolean = false
   protected var requiredSchema = StructType(dataSchema.fields ++ partitionSchema.fields)
 
   override def pruneColumns(requiredSchema: StructType): Unit = {
+    // [SPARK-30107] While the passed `requiredSchema` always have pruned nested columns, the actual
+    // data schema of this scan is determined in `readDataSchema`. File formats that don't support
+    // nested schema pruning, use `requiredSchema` as a reference and perform the pruning partially.
     this.requiredSchema = requiredSchema
```

Comment:
could we write "use `requiredSchema` as a reference and perform the pruning on top level."? 

BTW, for data source such as csv or json, even the top level pruning is not supported. In this case, does it mean the `readDaraSchema` will be always the full schema?

Reply:
Okay, I figure. For those data sources that don't support top level pruning, we will only return the required top level columns in readDataSchema. I guess in this case, the reader implementations still read the full data, and handle it internally, but pass less data into Spark. Wondering why we can not do similar thing in readers for nested data structure?

Reply:
I think you are right and it seems that data sources such as CSV and JSON try to simply ignore columns that are not needed in readers (e.g. spark.sql.csv.parser.columnPruning.enabled).

While CSV doesn't support nested data, JSON can potentially benefit from this. I haven't checked the JSON reader in detail to see whether it will need any changed. Sounds like a potential follow-up?

Reply:
I've updated the comment as well.

Reply:
It's not in the scope of this PR, but I feel we could always pass the pruned `requiredSchema` to the readers even for those not supporting any schema pruning. Thus, the benefit will be 1) move less data into Spark even the readers still require read the full data 2) The code will be consistent between handling nested schema pruning and top level pruning.

Reply:
I agree and it will also be trivial because the changes in this layer are minimal.